### ISSUE TITLE
[BUG] The loss of deepep and alltoall not all same

### ISSUE THREAD

**Issue by lk137095576** [NVIDIA/Megatron-LM#1603]
**Describe the bug**
I train  a moe model with deepep and alltoall，find the loss is not all same,  0.48%  MRE of two lineMegatron-LM 0.12.0 release .  show as fig: 

![Image](https://github.com/user-attachments/assets/76eef8aa-d4ce-4e3b-b21a-ad211fd9a657)

the test_token_dispatcher.py does not compare the result of two Dispatcher.

**Comment by Victarry** [comment]
`AlltoAllDispatcher` and `FlexDispatcher` will dispatch the same set of tokens to expert but in different orders, this can cause numerical difference due to the accumulation order.

Below is a test script for comparing the set of tokens with different dispatchers:
```python
import argparse
import torch

from megatron.core import parallel_state
from megatron.core.tensor_parallel.random import model_parallel_cuda_manual_seed
from megatron.core.models.gpt.moe_module_specs import get_moe_module_spec
from megatron.core.transformer.moe.moe_layer import MoELayer
from megatron.core.transformer.transformer_config import TransformerConfig

def initialize_distributed():
    parallel_state.destroy_model_parallel()

    torch.distributed.init_process_group(backend="nccl")
    torch.cuda.set_device(torch.distributed.get_rank() % torch.cuda.device_count())

def get_deepseek_v3_config(
    token_dispatcher_type,
    **kargs,
):
    EP = parallel_state.get_expert_model_parallel_world_size()
    TP = parallel_state.get_tensor_model_parallel_world_size()
    PP = parallel_state.get_pipeline_model_parallel_world_size()
    if token_dispatcher_type == "flex":
        moe_enable_deepep = True
    else:
        moe_enable_deepep = False
    deepseek_config = TransformerConfig(
        num_layers=61,
        num_moe_experts=256,
        # Hidden size
        hidden_size=7168,
        moe_ffn_hidden_size=2048,
        # moe_shared_expert_intermediate_size=3072,
        moe_shared_expert_intermediate_size=None,
        gated_linear_unit=True,
        add_bias_linear=False,
        # Token drop pad
        moe_pad_expert_input_to_capacity=False,
        # Routing
        moe_router_pre_softmax=False,
        moe_router_topk=8,
        moe_aux_loss_coeff=1e-3,
        moe_permute_fusion=True,
        # moe_router_topk_limited_devices=3,
        # moe_router_topk_scaling_factor=16.0,
        # Training
        moe_token_dispatcher_type=token_dispatcher_type,
        # Attention
        num_attention_heads=128,
        # Parallelism size
        expert_model_parallel_size=EP,
        tensor_model_parallel_size=TP,
        pipeline_model_parallel_size=PP,
        sequence_parallel=TP>1,
        # Node-limited routing
        moe_router_group_topk=4,
        moe_router_num_groups=8,
        moe_enable_deepep=moe_enable_deepep,
        moe_router_dtype="fp32",
    )
    if kargs:
        for k, v in kargs.items():
            if hasattr(deepseek_config, k):
                setattr(deepseek_config, k, v)
    return deepseek_config

def allclose_test(baseline, ref):
    if baseline.dtype == torch.float32:
        rtol=1e-5
        atol=1e-8
    else:
        rtol=1e-2
        atol=1e-3
    assert torch.allclose(baseline, ref, rtol=rtol, atol=atol), f"Baseline: {baseline}, Ref: {ref} max diff: {torch.max(torch.abs(baseline - ref))}"

def compare_forward_outputs(
    model_config_fn,
    dtype=torch.bfloat16,
    capacity_factor=None,
    pad_to_capacity=False,
):
    num_tokens = 64
    hidden_size = 16
    if dtype == torch.float32:
        rtol=1e-5
        atol=1e-8
    else:
        rtol=1e-2
        atol=1e-3

    moe_spec = get_moe_module_spec(
        use_te=True,
        num_experts=160,
        moe_grouped_gemm=True,
        moe_use_legacy_grouped_gemm=False,
    )
    config_with_a2a = model_config_fn(
        token_dispatcher_type="alltoall", 
        hidden_size=hidden_size,
        moe_expert_capacity_factor=capacity_factor,
        moe_pad_expert_input_to_capacity=pad_to_capacity,
    )
    moe_layer_baseline = MoELayer(
        config_with_a2a, moe_spec.submodules, layer_number=1
    ).to(dtype)

    config_with_unified = model_config_fn(
        token_dispatcher_type="flex",
        hidden_size=hidden_size,
        moe_expert_capacity_factor=capacity_factor,
        moe_pad_expert_input_to_capacity=pad_to_capacity,
    )
    moe_layer_unified = MoELayer(
        config_with_unified, moe_spec.submodules, layer_number=1
    ).to(dtype)

    moe_layer_unified.router.weight.data.copy_(moe_layer_baseline.router.weight)

    hidden_states_baseline = torch.randn(num_tokens, hidden_size, device="cuda", dtype=dtype).requires_grad_(True)
    hidden_states_unified = hidden_states_baseline.clone().detach().requires_grad_(True)

    probs_baseline, routing_map_baseline = moe_layer_baseline.router(hidden_states_baseline)
    probs_unified, routing_map_unified = moe_layer_unified.router(hidden_states_unified)

    assert torch.equal(routing_map_baseline, routing_map_unified), "Routing map are not equal"
    assert torch.allclose(probs_baseline, probs_unified), "Probs are not close"

    dispatched_input_a2a, tokens_per_expert_a2a, _ = (
        moe_layer_baseline.token_dispatcher.token_permutation(
            hidden_states_baseline, probs_baseline, routing_map_baseline
        )
    )

    # Since the probs multiplied have been moved to swiglu, we should simulate probs multiply here
    dispatched_input_a2a = dispatched_input_a2a / config_with_a2a.moe_router_topk
    # scale by TP size to simulate the fc2 output is parallelized by TP
    dispatched_input_a2a = dispatched_input_a2a / config_with_a2a.tensor_model_parallel_size

    restore_output_baseline, mlp_bias = (
        moe_layer_baseline.token_dispatcher.token_unpermutation(
            dispatched_input_a2a, None
        )
    )

    dispatched_input_unified, tokens_per_expert_unified, _ = (
        moe_layer_unified.token_dispatcher.token_permutation(
            hidden_states_unified, probs_unified, routing_map_unified
        )
    )
    # Since the probs multiplied have been moved to swiglu, we should simulate probs multiply here
    dispatched_input_unified = dispatched_input_unified / config_with_unified.moe_router_topk
    # scale by TP size to simulate the fc2 output is parallelized by TP
    dispatched_input_unified = dispatched_input_unified / config_with_unified.tensor_model_parallel_size

    restore_output_unified, mlp_bias = (
        moe_layer_unified.token_dispatcher.token_unpermutation(
            dispatched_input_unified, None
        )
    )


    random_grad = torch.randn_like(restore_output_baseline)
    restore_output_baseline.backward(random_grad)
    restore_output_unified.backward(random_grad)
    torch.cuda.synchronize()

    # 1. Test restored outputs are close to original hidden states
    assert torch.equal(
        tokens_per_expert_a2a, tokens_per_expert_unified
    ), f"Tokens per expert A2A: {tokens_per_expert_a2a}, Tokens per expert Unified: {tokens_per_expert_unified}"
    if capacity_factor is None:
        allclose_test(restore_output_baseline, hidden_states_baseline)
        allclose_test(restore_output_unified, hidden_states_unified)
    else:
        allclose_test(restore_output_baseline, restore_output_unified)

    # 2. Test tokens dispached to the same expert are the same
    tokens_per_expert = tokens_per_expert_a2a.cpu().tolist()
    cur_base = 0

    # Test if the dispatched tokens for each expert are the same
    assert (
        dispatched_input_a2a.shape == dispatched_input_unified.shape
    ), f"A2A dispatched input shape {dispatched_input_a2a.shape} does not match Unified dispatched input shape {dispatched_input_unified.shape}"
    for i, num_tokens in enumerate(tokens_per_expert):
        baseline = torch.sort(dispatched_input_a2a[cur_base : cur_base + num_tokens], dim=0)[0]
        ref = torch.sort(dispatched_input_unified[cur_base : cur_base + num_tokens], dim=0)[0]
        assert torch.equal(
            baseline, ref
        ), f"Baseline: {baseline}, Ref: {ref}"
        cur_base += num_tokens
    
    # 3. Test the backward gradient are all close
    allclose_test(hidden_states_baseline.grad, hidden_states_unified.grad)
    allclose_test(moe_layer_baseline.router.weight.grad, moe_layer_unified.router.weight.grad)

    torch.distributed.barrier()
    torch.cuda.synchronize()
    if torch.distributed.get_rank() == 0:
        print(f"✅ Accuracy test passed for flex dispatcher with dtype {dtype} and capacity factor {capacity_factor} and pad to capacity {pad_to_capacity}")
    torch.distributed.barrier()

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--tp-size", type=int, default=1, help="Tensor parallel size")
    parser.add_argument("--ep-size", type=int, default=1, help="Expert parallel size")
    parser.add_argument(
        "--model",
        type=str,
        choices=["deepseek", "mixtral"],
        default="deepseek",
        help="Model to benchmark (deepseek or mixtral)",
    )
    return parser.parse_args()


NUM_TOKENS = 4096
if __name__ == "__main__":

    # Megatron core distributed training initialization
    initialize_distributed()
    args = parse_args()
    EP = args.ep_size
    TP = args.tp_size
    PP = 1

    # Select model configuration based on argument
    model_config_fn = get_deepseek_v3_config

    parallel_state.initialize_model_parallel(
        TP,
        PP,
        expert_model_parallel_size=EP,
    )
    model_parallel_cuda_manual_seed(123)

    compare_forward_outputs(model_config_fn, dtype=torch.float32)
    compare_forward_outputs(model_config_fn, dtype=torch.bfloat16)
    parallel_state.destroy_model_parallel()
    torch.distributed.destroy_process_group()

```

You can run this script with:
```bash
torchrun --nproc_per_node=8 --nnodes=1 --master_port=29500 main.py --model deepseek --tp-size 1 --ep-size 8
torchrun --nproc_per_node=8 --nnodes=1 --master_port=29500 main.py --model deepseek --tp-size 2 --ep-size 4
```

**Comment by lk137095576** [comment]
@Victarry  Thank you for your explanation, I won't worry about it anymore. And from my test case, deepep converges faster, which should be a good thing.
